src/backend/oneapi/kernel/rotate.hpp

```diff
@@ -0,0 +1,167 @@
+};
+
+template<>
+struct itype_t<cfloat> {
+    typedef float wtype;
+    typedef cfloat vtype;
+};
+
+template<>
+struct itype_t<cdouble> {
+    typedef double wtype;
+    typedef cdouble vtype;
+};
+
+template<typename T, int dimensions>
+using local_accessor =
+    sycl::accessor<T, dimensions, sycl::access::mode::read_write,
+                   sycl::access::target::local>;
+
+template<typename T>
+void rotate(Param<T> out, Param<T> in, float theta, af_interp_type method,
+            int order) {
+    constexpr int TX = 4;
+    constexpr int TY = 4;
+    constexpr int TI = 4;  // Used for batching images
+    constexpr bool isComplex =
+        static_cast<af_dtype>(af::dtype_traits<T>::af_type) == c32 ||
+        static_cast<af_dtype>(af::dtype_traits<T>::af_type) == c64;
+
+    const float c = cos(-theta), s = sin(-theta);
+    float tx, ty;
+    {
+        const float nx = 0.5 * (in.info.dims[0] - 1);
+        const float ny = 0.5 * (in.info.dims[1] - 1);
+        const float mx = 0.5 * (out.info.dims[0] - 1);
+        const float my = 0.5 * (out.info.dims[1] - 1);
+        const float sx = (mx * c + my * -s);
+        const float sy = (mx * s + my * c);
+        tx             = -(sx - nx);
+        ty             = -(sy - ny);
+    }
+
+    const float tmat[6] = {
+        (float)round(c * 1000) / 1000.0f,  (float)round(-s * 1000) / 1000.0f,
+        (float)round(tx * 1000) / 1000.0f, (float)round(s * 1000) / 1000.0f,
+        (float)round(c * 1000) / 1000.0f,  (float)round(ty * 1000) / 1000.0f};
+
+    const auto local = sycl::range{TX, TY};
+
+    size_t nimages        = in.info.dims[2];
+    const size_t nbatches = in.info.dims[3];
+
+    size_t global_x              = local[0] * divup(out.info.dims[0], local[0]);
+    size_t global_y              = local[1] * divup(out.info.dims[1], local[1]);
+    const size_t blocksXPerImage = global_x / local[0];
+    const size_t blocksYPerImage = global_y / local[1];
+
+    if (nimages > TI) {
+        const size_t tile_images = divup(nimages, TI);
+        nimages                  = TI;
+        global_x                 = global_x * tile_images;
+    }
+    global_y *= nbatches;
+
+    const auto global = sycl::range{global_x, global_y};
+
+    getQueue().submit([&](sycl::handler &h) {
+        auto d_out = out.data->get_access(h);
+        auto d_in  = in.data->get_access(h);
+
+        h.parallel_for(sycl::nd_range{global, local}, [=](sycl::nd_item<2> it) {
+            sycl::group g = it.get_group();
+
+            // Compute which image set
+            const int setId      = g.get_group_id(0) / blocksXPerImage;
+            const int blockIdx_x = g.get_group_id(0) - setId * blocksXPerImage;
+
+            const int batch      = g.get_group_id(1) / blocksYPerImage;
+            const int blockIdx_y = g.get_group_id(1) - batch * blocksYPerImage;
+
+            // Get thread indices
+            const int xido = it.get_global_id(0);
+            const int yido = it.get_global_id(1);
```

Comment:
revert to blockIdx_x based indexing. will investigate more after interp.hpp is integrated